File: old/ivas-go/internal/domain/repository/rubric_repository.go
package repository

import (
	"github.com/ivas/internal/domain/entity"
	"gorm.io/gorm"
)

// RubricRepository handles rubric database operations
type RubricRepository struct {
	db *gorm.DB
}

// NewRubricRepository creates a new rubric repository
func NewRubricRepository(db *gorm.DB) *RubricRepository {
	return &RubricRepository{db: db}
}

// Create creates a new rubric
func (r *RubricRepository) Create(rubric *entity.Rubric) error {
	return r.db.Create(rubric).Error
}

// CreateBatch creates multiple rubrics
func (r *RubricRepository) CreateBatch(rubrics []entity.Rubric) error {
	return r.db.Create(&rubrics).Error
}

// FindByQuestionID finds rubric by question ID
func (r *RubricRepository) FindByQuestionID(questionID string) (*entity.Rubric, error) {
	var rubric entity.Rubric
	err := r.db.First(&rubric, "question_id = ?", questionID).Error
	if err != nil {
		return nil, err
	}
	return &rubric, nil
}

// Update updates a rubric
func (r *RubricRepository) Update(rubric *entity.Rubric) error {
	return r.db.Save(rubric).Error
}

File: app/repositories/rubric_repository.py
"""
Rubric Repository

Data access layer for Rubric models. Handles operations for creating and updating
grading rubrics associated with questions. Called by the QuestionService when
new questions are generated along with their rubrics.
"""

from typing import Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.question import Rubric


class RubricRepository:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def create(self, rubric: Rubric) -> Rubric:
        self.session.add(rubric)
        await self.session.flush()
        await self.session.refresh(rubric)
        return rubric

    async def create_batch(self, rubrics: list[Rubric]) -> list[Rubric]:
        self.session.add_all(rubrics)
        await self.session.flush()
        return rubrics

    async def find_by_question_id(self, question_id: str) -> Optional[Rubric]:
        result = await self.session.execute(
            select(Rubric).where(Rubric.question_id == question_id)
        )
        return result.scalar_one_or_none()

    async def update(self, rubric: Rubric) -> Rubric:
        await self.session.merge(rubric)
        await self.session.flush()
        return rubric
